payroll-processor-client/src/app/payroll/payroll-charts/payroll-charts.component.ts

```diff
@@ -0,0 +1,83 @@
+import {
+  Component,
+  OnInit,
+} from '@angular/core';
+
+import { DataService } from 'src/app/data/data-service';
+import { NameValue } from 'src/app/data/state/data-model';
+
+@Component({
+  selector: 'app-payroll-charts',
+  templateUrl: './payroll-charts.component.html',
+  styleUrls: ['./payroll-charts.component.scss'],
+})
+export class PayrollChartsComponent implements OnInit {
+  departmentDistributionData: NameValue[] = [];
+  employeeDistributionData: NameValue[] = [];
+  view: [number, number] = [700, 400];
+
+  // options+
+  gradient = false;
+  showLegend = true;
+  showLabels = false;
+  isDoughnut = false;
+  legendPosition = 'below';
+
+  colorScheme = {
+    domain: ['#5AA454', '#A10A28', '#C7B42C', '#4287f5', '#AAAAAA'],
+  };
+
+  constructor(private dataService: DataService) {
+    this.prepData();
+  }
+
+  ngOnInit(): void {}
+
+  prepData() {
+    this.prepDepartmentDistributionData();
+    this.prepEmployeeDistributionData();
+  }
+
+  prepEmployeeDistributionData() {
+    const payroll = this.dataService.getPayroll();
+    const employees = this.dataService.getEmployees();
+    const map: Map<string, number> = new Map();
+
+    payroll.forEach((p) => {
+      if (p.employeeStatus !== 'ACTIVE') {
+        return;
+      }
+      return map.has(p.employeeName)
+        ? map.set(p.employeeName, map.get(p.employeeName) + p.grossPayroll)
+        : map.set(p.employeeName, p.grossPayroll);
+    });
+
+    this.employeeDistributionData = Array.from(
+      map,
+      ([key, value]) => new NameValue(key, value),
+    );
+  }
+
+  prepDepartmentDistributionData() {
```

Comment:
Without getting into too much architecture (rxjs, akita, ect...), we could at least move these data processing operations out of the component and into a "Presenter".

Think of a Presenter as a service class that is directly tied to this component.

Components are glue between state and HTML, any other processing should be put somewhere else if possible.

To make this change you can create a new service class `payroll-charts.presenter.ts` and inject it into this component.

Give it 2 methods `prepDepartmentDistributionData`, and `prepEmployeeDistributionData` that return your `Map`s.

Call those methods in this component's constructor to assign to the component `Map` fields.